File: src/main/scala/experiments/codingChallenges/FizzBuzz.scala
package experiments.codingChallenges

/**
 * inspired by https://www.youtube.com/watch?v=QPZ0pIK_wsc
 */
object FizzBuzz extends App {

  val specials = Map(
    3 -> "Fizz",
    5 -> "Buzz"
  ).map((special _).tupled)

  (1 to 100).map(convert).foreach(println)

  def convert(i: Int): String = {
    (specials :\ "")(_(i) + _) match {
      case "" => i.toString
      case s => s
    }
  }

  def special(divider: Int, name: String): Int => String = {
    i => if (i % divider == 0) name else ""
  }
}

File: src/main/haskell/FizzBuzz.hs
module FizzBuzz where

main = mapM_ print (fmap convert [1..100])

convert :: Int -> String
convert x | string == "" = show x
          | otherwise    = string
    where string = foldr (\f -> \y -> f(x) ++ y) "" specials
          specials = (special 3 "Fizz") : (special 5 "Buzz") : []

special :: Int -> String -> Int -> String
special divider name = \x -> if (x `mod` divider == 0) then name else ""

